### DISCUSSION TITLE
Investigation of fix for Do not fail Maven Central Rules sources or javadocs jar missing for an empty jar.

### DISCUSSION THREAD

**Discussion by ge0ffrey** [jreleaser/jreleaser#1355] [Q&A] [answered]
https://github.com/jreleaser/jreleaser/issues/1289 is fixed,
and I am testing that fix,
but **my test for it fails**.

I am using this thread to figure out where the issue lies in the JReleaser source code.

**Comment by ge0ffrey** [comment]
AbstractMavenDeployer.requiresSourcesJar investigation

Here's the commit that fixes the original issue
https://github.com/jreleaser/jreleaser/commit/10742852a58603b852c6debafb9d50034d956b58#diff-9b8678cd528d74dcc05e655f431d1ac1c3ff0a061df7b811e82faa5126a28bf4R231

**Reply by ge0ffrey** [reply]
For an empty jar (such as foo-spring-boot-starter which has 0 code, but depends on foo-autoconfigure and all that's optional dependencies)
this seems to return true inAbstractMavenDeployer, while it should return false:

```
    private boolean requiresSourcesJar(Deployable deployable) {
        if (!deployable.requiresSourcesJar()) return false;

        Optional<org.jreleaser.model.internal.deploy.maven.MavenDeployer.ArtifactOverride> override = getDeployer().getArtifactOverrides().stream()
            .filter(a -> a.getGroupId().equals(deployable.getGroupId()) && a.getArtifactId().equals(deployable.getArtifactId()))
            .findFirst();

        if (override.isPresent() && (override.get().isSourceJarSet())) return override.get().isSourceJar();

        return getDeployer().isSourceJar();
    }
```

**Reply by aalmiray** [reply]
Did you try defining an explicit artifact override to exclude this artifact?

I know we discussed an option for JReleaser to inspect JARs to determine if sources and javadoc would be needed but in the end it was not added. Sonatype's documentation clearly states that sources and javadoc are needed. Yet, they seem to be running additional, undocumented rules. 

For this reason it's not feasible to reproduce these rules without being 100% sure it was done correctly.

The compromise is to define explicit artifact overrides. 

If after defining said overrides you still get a failure during deployment then it would mean there's a bug in the code. 

Please run deploy in dryrun mode to find out.

**Reply by ge0ffrey** [reply]
Yes, you're right about those secret rules.

Because if I hack JReleaser like this:

```
    private boolean requiresSourcesJar(Deployable deployable) {
        return false;
    }

    private boolean requiresJavadocJar(Deployable deployable) {
        return false;
    }
```

then I can close the staging repository just fine - and release it:

![image](https://user-images.githubusercontent.com/176880/233827165-242b4761-8669-4b39-8a84-7a6ed037dece.png)

**Reply by aalmiray** [reply, marked as answer]
Did you try using artifact overrides as explained in the doc? There should not be a need to hack JReleaser to make this work. If there is then a bug is still lurking.

**Reply by ge0ffrey** [reply]
No, I didn't. I wasn't aware about "artifact overrides".

I first looked for it in the early-access online docs, both through the search function and finding the word "override" on this page:
  https://jreleaser.org/guide/early-access/examples/maven/maven-central.html
but no luck that way.

Then I found it here:
  https://jreleaser.org/guide/early-access/reference/deploy/maven/nexus2.html
I'll try that out.

Thank you very much, Andres!

**Reply by ge0ffrey** [reply]
This works like a charm:

```
          <jreleaser>
            ...
            <deploy>
              <maven>
                <nexus2>
                  <maven-central>
                    ...
                    <artifactOverrides>
                      <artifactOverride>
                        <groupId>org.foo</groupId>
                        <artifactId>foo-spring-boot-starter</artifactId>
                        <sourceJar>false</sourceJar>
                        <javadocJar>false</javadocJar>
                      </artifactOverride>
                    </artifactOverrides>
                  </maven-central>
                </nexus2>
              </maven>
            </deploy>
          </jreleaser>
```
